src/tests/ftest/util/apricot/apricot/test.py

```diff
@@ -124,61 +129,95 @@ def setUp(self):
         self.d_log = DaosLog(self.context)
 
     def tearDown(self):
+        """Tear down after each test case."""
         super(TestWithoutServers, self).tearDown()
 
         if self.fault_file:
             os.remove(self.fault_file)
 
+
 class TestWithServers(TestWithoutServers):
-    '''
-    Run tests with DAOS servers
+    """Run tests with DAOS servers and at least one client.
+
+    Optionally run DAOS clients on specified hosts.  By default run a single
+    DAOS client on the host executing the test.
 
     :avocado: recursive
-    '''
+    """
+
     def __init__(self, *args, **kwargs):
+        """Initialize a TestWithServers object."""
         super(TestWithServers, self).__init__(*args, **kwargs)
 
         self.agent_sessions = None
 
     def setUp(self):
+        """Set up each test case."""
         super(TestWithServers, self).setUp()
 
-        self.server_group = self.params.get("server_group", '/server/',
-                                            'daos_server')
-        self.hostlist_servers = self.params.get("test_machines", '/run/hosts/*')
-
+        self.server_group = self.params.get(
+            "server_group", "/server/", "daos_server")
+
+        # Determine which hosts to use as servers and optionally clients.
+        # Support the use of a host type count to test with subsets of the
+        # specified hosts lists
+        test_machines = self.params.get("test_machines", "/run/hosts/*")
+        test_servers = self.params.get("test_servers", "/run/hosts/*")
+        test_clients = self.params.get("test_clients", "/run/hosts/*")
+        server_count = self.params.get("server_count", "/run/hosts/*")
+        client_count = self.params.get("client_count", "/run/hosts/*")
+
+        # Supported combinations of yaml hosts arguments:
+        #   - test_machines [+ server_count]
+        #   - test_servers [+ server_count]
+        #   - test_servers [+ server_count] + test_clients [+ client_count]
+        if test_machines:
+            self.hostlist_servers = test_machines[:server_count]
+        elif test_servers and test_clients:
+            self.hostlist_servers = test_servers[:server_count]
+            self.hostlist_clients = test_clients[:client_count]
+        elif test_servers:
+            self.hostlist_servers = test_servers[:server_count]
+        self.log.info("hostlist_servers:  {}".format(self.hostlist_servers))
```

Comment:
(pylint-logging-format-interpolation) Use % formatting in logging functions and pass the % parameters as arguments